notype when inlining pattern matching

```scala
object Test {
  inline def openImpl(n: Some[Int]): Int =
    n match { case Some(i) => i }

  def open(a: Some[Int]): Int = openImpl(a)
}
```

<details>
  <summary>Stack trace</summary>
<pre>
Exception in thread "main" java.lang.AssertionError: assertion failed: <notype> class dotty.tools.dotc.ast.Trees$TypeTree 115
        at scala.Predef$.assert(Predef.scala:219)
        at dotty.tools.dotc.typer.ReTyper.promote(ReTyper.scala:29)
        at dotty.tools.dotc.typer.ReTyper.typedTypeTree(ReTyper.scala:52)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1652)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:1753)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedDefDef$1(Typer.scala:1294)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1287)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1620)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1678)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1716)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1739)
        at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:620)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedBlock$1(Typer.scala:623)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:622)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1639)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedFunction$1(Typer.scala:846)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:693)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1641)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2108)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2275)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1855)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1851)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1851)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1751)
        at dotty.tools.dotc.typer.ReTyper.typedUnApply(ReTyper.scala:61)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1665)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.typedPattern(Typer.scala:1755)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedCase$1(Typer.scala:964)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:918)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedCases$3(Typer.scala:912)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:62)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:78)
        at dotty.tools.dotc.typer.Typer.typedCases(Typer.scala:912)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedMatch$1(Typer.scala:884)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedMatch(Typer.scala:875)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1644)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:94)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Inliner.$anonfun$inlined$8(Inliner.scala:475)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:140)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Inliner.inlined(Inliner.scala:472)
        at dotty.tools.dotc.typer.Inliner$.inlineCall(Inliner.scala:245)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2146)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2275)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1855)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1851)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1851)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1751)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedDefDef$1(Typer.scala:1305)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1287)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1620)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1678)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1716)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1739)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedClassDef$1(Typer.scala:1405)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1335)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1623)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1678)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1716)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1739)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedPackageDef$1(Typer.scala:1522)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1509)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1662)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1679)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1694)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented(Reporter.scala:169)
        at dotty.tools.dotc.reporting.Reporting.doTraceIndented$(Reporter.scala:154)
        at dotty.tools.dotc.core.Contexts$Context.doTraceIndented(Contexts.scala:58)
        at dotty.tools.dotc.reporting.Reporting.dotty$tools$dotc$reporting$Reporting$$doTraceIndented(Reporter.scala:151)
        at dotty.tools.dotc.reporting.Reporting.traceIndented(Reporter.scala:141)
        at dotty.tools.dotc.reporting.Reporting.traceIndented$(Reporter.scala:139)
        at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:58)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1692)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1751)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$typeCheck$1(FrontEnd.scala:64)
        at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:32)
        at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:62)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$5(FrontEnd.scala:93)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$5$adapted(FrontEnd.scala:93)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:93)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:123)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:120)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:120)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:108)
        at dotty.tools.dotc.Run.compileSources(Run.scala:95)
        at dotty.tools.dotc.Run.compile(Run.scala:79)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
        at dotty.tools.dotc.Driver.process(Driver.scala:124)
        at dotty.tools.dotc.Driver.process(Driver.scala:93)
        at dotty.tools.dotc.Driver.process(Driver.scala:105)
        at dotty.tools.dotc.Driver.main(Driver.scala:132)
        at dotty.tools.dotc.Main.main(Main.scala)
java.lang.RuntimeException: Nonzero exit code returned from runner: 1
        at scala.sys.package$.error(package.scala:27)
</pre>
</details>

--- a/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
@@ -58,7 +58,9 @@ class ReTyper extends Typer {
   }
 
   override def typedUnApply(tree: untpd.UnApply, selType: Type)(implicit ctx: Context): UnApply = {
-    val fun1 = typedExpr(tree.fun, AnyFunctionProto)
+    val funProto = new UnapplyFunProto(WildcardType, this)
+
+    val fun1 = typedExpr(tree.fun, funProto)
     val implicits1 = tree.implicits.map(typedExpr(_))
     val patterns1 = tree.patterns.mapconserve(pat => typed(pat, pat.tpe))
     untpd.cpy.UnApply(tree)(fun1, implicits1, patterns1).withType(tree.tpe)

--- a/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
@@ -58,9 +58,7 @@ class ReTyper extends Typer {
   }
 
   override def typedUnApply(tree: untpd.UnApply, selType: Type)(implicit ctx: Context): UnApply = {
-    val funProto = new UnapplyFunProto(WildcardType, this)
-
-    val fun1 = typedExpr(tree.fun, funProto)
+    val fun1 = typedUnadapted(tree.fun, AnyFunctionProto)
     val implicits1 = tree.implicits.map(typedExpr(_))
     val patterns1 = tree.patterns.mapconserve(pat => typed(pat, pat.tpe))
     untpd.cpy.UnApply(tree)(fun1, implicits1, patterns1).withType(tree.tpe)

--- a/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ReTyper.scala
@@ -104,4 +104,6 @@ class ReTyper extends Typer {
     Implicits.NoImplicitMatches
   override def checkCanEqual(ltp: Type, rtp: Type, pos: Position)(implicit ctx: Context): Unit = ()
   override def inlineExpansion(mdef: DefDef)(implicit ctx: Context): List[Tree] = mdef :: Nil
+
+  override protected def checkEqualityEvidence(tree: tpd.Tree, pt: Type)(implicit ctx: Context): Unit = ()
 }

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -3,7 +3,7 @@ package dotc
 package typer
 
 import core._
-import ast._
+import ast.{tpd, _}
 import Trees._
 import Constants._
 import StdNames._
@@ -28,15 +28,17 @@ import EtaExpansion.etaExpand
 import dotty.tools.dotc.transform.Erasure.Boxing
 import util.Positions._
 import util.common._
-import util.{SourcePosition, Property}
+import util.{Property, SourcePosition}
+
 import collection.mutable
 import annotation.tailrec
 import Implicits._
-import util.Stats.{track, record}
-import config.Printers.{typr, gadts}
+import util.Stats.{record, track}
+import config.Printers.{gadts, typr}
 import rewrite.Rewrites.patch
 import NavigateAST._
 import transform.SymUtils._
+
 import language.implicitConversions
 import printing.SyntaxHighlighting._
 
@@ -2122,18 +2124,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
           typed(untpd.Select(untpd.TypedSplice(tree), nme.apply), pt)
         }
         else if (ctx.mode is Mode.Pattern) {
-          tree match {
-            case _: RefTree | _: Literal
-            if !isVarPattern(tree) &&
-               !(tree.tpe <:< pt)(ctx.addMode(Mode.GADTflexible)) &&
-               !this.isInstanceOf[ReTyper] =>
-              val cmp =
-                untpd.Apply(
-                  untpd.Select(untpd.TypedSplice(tree), nme.EQ),
-                  untpd.TypedSplice(dummyTreeOfType(pt)))
-              typedExpr(cmp, defn.BooleanType)(ctx.retractMode(Mode.Pattern))
-            case _ =>
-          }
+          checkEqualityEvidence(tree, pt)
           tree
         }
         else if (tree.tpe <:< pt) {
@@ -2276,6 +2267,20 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
               else adaptNoArgs(wtp)
           }
       }
+    }
+  }
+
+  protected def checkEqualityEvidence(tree: tpd.Tree, pt: Type)(implicit ctx: Context) : Unit = {
+    tree match {
+      case _: RefTree | _: Literal
+        if !isVarPattern(tree) &&
+          !(tree.tpe <:< pt) (ctx.addMode(Mode.GADTflexible)) =>
+        val cmp =
+          untpd.Apply(
+            untpd.Select(untpd.TypedSplice(tree), nme.EQ),
+            untpd.TypedSplice(dummyTreeOfType(pt)))
+        typedExpr(cmp, defn.BooleanType)(ctx.retractMode(Mode.Pattern))
+      case _ =>
     }
   }
 }

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -2125,7 +2125,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
           tree match {
             case _: RefTree | _: Literal
             if !isVarPattern(tree) &&
-               !(tree.tpe <:< pt)(ctx.addMode(Mode.GADTflexible)) =>
+               !(tree.tpe <:< pt)(ctx.addMode(Mode.GADTflexible)) &&
+               !this.isInstanceOf[ReTyper] =>
               val cmp =
                 untpd.Apply(
                   untpd.Select(untpd.TypedSplice(tree), nme.EQ),

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -2270,6 +2270,12 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     }
   }
 
+  /** Check that `tree == x: pt` is typeable. Used when checking a pattern
+    * against a selector of type `pt`. This implementation accounts for
+    * user-defined definitions of `==`.
+    *
+    * Overwritten to no-op in ReTyper.
+    */
   protected def checkEqualityEvidence(tree: tpd.Tree, pt: Type)(implicit ctx: Context) : Unit = {
     tree match {
       case _: RefTree | _: Literal
@@ -2279,7 +2285,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
           untpd.Apply(
             untpd.Select(untpd.TypedSplice(tree), nme.EQ),
             untpd.TypedSplice(dummyTreeOfType(pt)))
-        typedExpr(cmp, defn.BooleanType)(ctx.retractMode(Mode.Pattern))
+        typedExpr(cmp, defn.BooleanType)
       case _ =>
     }
   }

--- a/tests/pos/i3050.scala
+++ b/tests/pos/i3050.scala
@@ -1,0 +1,11 @@
+object Test {
+  inline def openImpl(): Int =
+    Some(42) match { case Some(i) => i }
+
+  def open(): Int = openImpl()
+
+  inline def openImpl2(): Int =
+    Some(42) match { case None => 42 }
+
+  def open2(): Int = openImpl2()
+}

